### PR TITLE
PUBDEV-5249: add first batch of Runit tests ported from GBM to XGBoost and fixes for bugs found

### PR DESCRIPTION
**Pull request by rpeck**

### CODE REVIEW

**Comment by Pscheidl** [comment]
Rebased on top of latest master as well.

**Comment by Pscheidl** [comment]
The test `runit_XGBoostRandomGrid_airlines_large.R` which is failing may reveal a possible bug.

**Comment by Pscheidl** [comment]
Tests to fix:
- runit_XGBoostRandomGrid_airlines_large.R
- unit_XGBoostGrid_airlines.R
- runit_XGBoost_KDD_trees_large.R

**Comment by Pscheidl** [comment]
Using original domains was breaking XGBoost, as it can not work with strings of course.

**Comment by Pscheidl** [comment]
R test fixed, yet still failing:
- runit_XGBoostRandomGrid_airlines_large.R (has been revised & fixed for a few issues, possibly points to an internal issue inside XGBoost integration)
- runit_XGBoost_KDD_trees_large.R

**Comment by Pscheidl** [comment]
Python fails:

```plain
Test Result (3 failures / -14)

    Py3.6 Test Demos / r_suite.pyunit_lending_club_munging_assembly_large.py
    R3.4 Demos Small / r_suite.h2o.deeplearning.R
    Py3.5 Small / r_suite.pyunit_automl_leaderboard.py
```

File: h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java [hunk 1]
```diff
@@ -64,24 +64,24 @@ public boolean equals(Object o) {
 
       if (vecs.length != fm.vecs.length) {
         Log.warn("Training frame vec count has changed from: " +
-                fm.vecs.length + " to: " + fm.vecs.length);
+                vecs.length + " to: " + fm.vecs.length);
         error = true;
       }
       if (fm.names.length != fm.names.length) {
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java]: this is always false

File: h2o-r/tests/testdir_algos/xgboost/runit_XGBoostRandomGrid_airlines_large.R [hunk 2]
```diff
@@ -40,7 +40,7 @@ xgboost.random.grid.test <- function() {
       ## Not available in XGBoost:      histogram_type = c("UniformAdaptive","QuantilesGlobal","RoundRobin")
 
       ## Some XGBoost-specific parameters:
-      tree_method = c("auto", "exact", "approx", "hist"),
+      tree_method = c("auto", "exact", "hist", "approx"),
```
Comment (Pscheidl) [inline h2o-r/tests/testdir_algos/xgboost/runit_XGBoostRandomGrid_airlines_large.R]: lossguide requires hist

Comment (Pscheidl) [reply h2o-r/tests/testdir_algos/xgboost/runit_XGBoostRandomGrid_airlines_large.R]: There are combinations, won't work anyway.

File: h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java (hunk 1 above)
Comment (Pscheidl) [reply h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java]: fixed, removed the fm

**Comment by Pscheidl** [comment]
Tests failed are unrelated:

```plain
    # Artificial data to be used throughout the test, very simple
    train_frame = pd.read_csv(pyunit_utils.locate('smalldata/testng/airquality_train1.csv'))

    data = train_frame.iloc[:, 1:];
    label = train_frame.iloc[:,:1]

    # Native XGBosot model trained first
    model = xgb.XGBClassifier(random_state=0, colsample_bytree = 0.3)
    model.fit(data, label.as_matrix().flatten())
    preds = model.predict(data)
    print(preds)
```

**Comment by Pscheidl** [comment]
We seem to hit the following error:
https://github.com/dmlc/rabit/blob/master/src/socket.h#L511

tils::Assert(maxfd < FD_SETSIZE, "maxdf must be smaller than FDSETSIZE");

**Comment by Pscheidl** [comment]
The socket leak has been found on two places. Misallocated `ServerSocketChannel.open()` may thrown an exception when tried to be bound, but those connections seem to create hanging FDs until closed or garbage collected (or the process is terminated).

Also, if `SHUTDOWN_CMD` is received, there is no need to keep the connection opened, thus closing it immediately seems like a best option to free resources ASAP.

Last but not least, the thread where `final SocketChannel channel = tracker.sock.accept();` is called (in its run method) may be interrupted (by calling `this.trackerThread.interrupt();` from `RabitTrackerH2O`'s `stop()` method. The `ServerSocketChannel` itself is closed explicitely, but not the accepted connections. If the thread is interrupted, then all the SocketChannels are closed explicitely. Calling `close()` method one more time (in case it had been called before as a reaction to `SHUTDOWN_CMD`) does no harm, as java.nio first checks for the connection being open before an attempt to close it is made.

Rebased on top of master (there was a big update to `XGBoostUpdateTask`)

**Comment by Pscheidl** [comment]
Won't work until rebased on top of:
https://github.com/h2oai/h2o-3/pull/2347

(tested before in this PR, we decided to fix the Rabit FD leak in a separate issue with @michalkurka )

**Comment by Pscheidl** [comment]
Rebased on top of latest master, where #2347 is present already. Witing for tests, then merging.

**Comment by Pscheidl** [comment]
Test failure is related to AutoML test, not XGBoost. Merging.

**Pull request merged**